Suggest using explicit deriving strategies for newtypes

divarvel opened the pull request:
Following #682, this is a first shot at suggesting explicit deriving strategies.

There are still a few questions before considering this patch satisfying

- ~how do I make it opt-in? (it makes a newtype-related test fail)~
- for the lack of a better solution, it always suggests `deriving newtype`, that could change behaviour (I don't think i can compute the actual, implicit behaviour)
- it covers gadt-like newtype definitions, I'm not 100% sure of the implications
- ~`NewTypeDerivingStrategies` is quite a long name and it's tedious for vertical alignment. Maybe you'd prefer a shorter name?~

divarvel commented:
@ndmitchell The only blocking issue IMO is about making the hint optional (it is the cause for the test failure)

ndmitchell commented on src/Hint/NewTypeDerivingStrategies.hs:
```diff
@@ -0,0 +1,38 @@
+
+import Hint.Type
+
+newTypeDerivingStrategiesHint :: DeclHint
```
I'd just shove this in the Newtype package of hints - it's close enough already and not that long.

divarvel replied:
Done

ndmitchell commented on src/Hint/NewTypeDerivingStrategies.hs:
```diff
@@ -0,0 +1,38 @@
+newtypeDecl :: Decl_ -> Maybe Decl_
+newtypeDecl (DataDecl x1 x2@(NewType _) x3 x4 x5 x6)
+    | any hasNoStrategy x6 = Just $ DataDecl x1 x2 x3 x4 x5 (withNewtype <$> x6)
+
```
Please delete this blank line - otherwise it starts to look like a partial pattern match.

divarvel replied:
indeed, good catch

ndmitchell commented on src/Hint/NewTypeDerivingStrategies.hs:
```diff
@@ -0,0 +1,38 @@
+newTypeDerivingStrategiesHint _ _ = newTypeDerivingStrategiesHintDecl
+
+newTypeDerivingStrategiesHintDecl :: Decl_ -> [Idea]
+newTypeDerivingStrategiesHintDecl x
```
```
newTypeDerivingStrategiesHintDecl x = [suggestN "Use DerivingStrategies" x new | Just new <- [newtypeDecl x]]
```

ndmitchell commented on src/Hint/NewTypeDerivingStrategies.hs:
```diff
@@ -0,0 +1,38 @@
+newTypeDerivingStrategiesHintDecl :: Decl_ -> [Idea]
+newTypeDerivingStrategiesHintDecl x
+    | Just new <- newtypeDecl x
+    = [suggestN "Use DerivingStrategies" x new
```
Instead of `suggestN` you need to define `ignoreN`.

divarvel replied:
:ok_hand:

divarvel commented:
I've applied all of your suggestions, added a `ignoreN` in a separate commit and moved the code to `Hint.NewType`. There are now test failures because the expected rewrite is not proposed, plus a missing file error I'm not sure I understand.

ndmitchell commented on src/Hint/All.hs:
```diff
@@ -38,29 +38,30 @@ data HintBuiltin =
     HintBracket | HintNaming | HintPattern | HintImport | HintExport |
     HintPragma | HintExtensions | HintUnsafe | HintDuplicate | HintRestrict |
-    HintComment | HintNewType | HintSmell
+    HintComment | HintNewType | HintSmell | HintNewTypeDerivingStrategies
```
I actually mean merge it even further so they are just under `HintNewtype` - no entry here at all.

divarvel replied:
:ok_hand:

ndmitchell commented on src/Hint/NewType.hs:
```diff
@@ -19,9 +19,12 @@ data A = A {b :: !C} -- newtype A = A {b :: C}
 data A = A Int#
 {-# LANGUAGE UnboxedTuples #-}; data WithAnn x = WithAnn (# Ann, x #)
 data A = A () -- newtype A = A ()
+newtype Foo = Foo Int deriving (Show, Eq) -- newtype Foo = Foo Int deriving newtype (Show, Eq)
```
I think `-- @Ignore` as a prefix might make the test pass

divarvel replied:
It's not needed, the tests pass without the `@Ignore` annotation

divarvel commented:
It looks like tests are green! Last question, how can I try it on a real project?

ndmitchell commented:
Yep, if @ignore is not required then no need to put it in, and that is indeed how you test it.

ndmitchell commented:
Thanks! Looks great.

divarvel commented:
Thanks for your help!

ndmitchell commented:
No problem - planning to make a release once #681 is fixed.
